### PR TITLE
Create TaskDetail E-mail

### PR DESCRIPTION
**Pull request by gburnhill**

### CODE REVIEW

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        private readonly IConnectRequestService _connectRequestService;
+        private readonly IOptions<EmailConfig> _emailConfig;
+        private readonly ICosmosDbService _cosmosDbService;
+
+        List<SendMessageRequest> _sendMessageRequests;
+
+        public string GetUnsubscriptionGroupName(int? recipientId)
+        {
+
+                return UnsubscribeGroupName.TaskDetail;
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:39]: These emails are only generated by actions on behalf of the user they will be sent to.  I wonder if there's a way we can make them unsubscribable...?

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            if (user == null)
+            {
+                throw new BadRequestException($"unable to retrieve user object for {recipientUserId.Value}");
+            }
+
+            var job = await _connectRequestService.GetJobDetailsAsync(jobId.Value);
+            var groupName = "";
+            var volunteerInstructions = "";
+
+            if (groupId.HasValue) {
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:75]: Should this be the ReferringGroupId from the job?  (And on lines 76 and 79.)

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                var group = await _connectGroupService.GetGroup(groupId.Value);
+                if (group != null)
+                groupName = group.Group.GroupName;
+                var instructions = await _connectGroupService.GetGroupSupportActivityInstructions(groupId.Value, job.JobSummary.SupportActivity);
+                
+                volunteerInstructions = $"{instructions.Intro} <br />";
+                if (instructions.Steps.Count > 0) {
+                    var steps = string.Join("", instructions.Steps.Select(x => $"<li>{x.Heading}: {x.Detail}</li>"));
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:83]: The Details are stored using Markdown, which should then be parsed into HTML.  In the Front End we're using the Westwind.AspNetCore.Markdown NuGet package to do that.  See _GroupSupportActivityInstructions.cshtml, although the syntax will be different when it's not in MVC.

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            var groupName = "";
+            var volunteerInstructions = "";
+
+            if (groupId.HasValue) {
+                var group = await _connectGroupService.GetGroup(groupId.Value);
+                if (group != null)
+                groupName = group.Group.GroupName;
+                var instructions = await _connectGroupService.GetGroupSupportActivityInstructions(groupId.Value, job.JobSummary.SupportActivity);
+                
+                volunteerInstructions = $"{instructions.Intro} <br />";
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:81]: As a style point, `<br />`s are not particularly desirable.  Better where possible to use margins.  (The principle is that html elements should have a meaning - and styling should be separate.)

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            var lineBreakRegex = new Regex(@"(\n|\r){1,2}");
+            var allQuestions = job.JobSummary.Questions.Where(q => q.ShowOnTaskManagement(false) && !string.IsNullOrEmpty(q.Answer));
+            var shopping = lineBreakRegex.Split(allQuestions.Where(x => x.Id == (int)Questions.Shopping_List).FirstOrDefault().Answer).ToList();
+            var shoppingList = string.Join("", shopping.Select(x => $"<tr><td>{x}</td></tr>"));
+            var otherQuestions = allQuestions.Where(x => x.Id != (int)Questions.Shopping_List);
+
+            var otherQuestionsList = string.Join("", otherQuestions.Select(x => $"<tr><td>{x.FriendlyName()}: </td></tr>" +
+            $"{string.Join("",lineBreakRegex.Split(x.Answer).Select(y => $"<tr><td{y}</td></tr>"))}"));
+
+
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:97]: Do we need to treat shopping lists differently from other questions here?  In the front end, we simply pass all answers through ToHtmlSafeStringWithLineBreaks(), which swaps new line charachters for *ahem* `<br />`s, and will also santisise any undesirable html users may have entered.  `<script>...`

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        public async  Task<List<SendMessageRequest>> IdentifyRecipients(int? recipientUserId, int? jobId, int? groupId, Dictionary<string, string> additionalParameters)
+        {
+
+            _sendMessageRequests.Add(new SendMessageRequest()
+            {
+                TemplateName = TemplateName.TaskDetail,
+                RecipientUserID = recipientUserId.Value,
+                GroupID = groupId,
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:122]: I may be wrong, but I don't think groupId is going to have any meaning here - so maybe don't pass it through?

File: CommunicationService/CommunicationService.SendGridManagement/Emails/TaskDetail.html [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿<p>Hi,</p>
+
+<p>Please see the details you requested below:</p>
+
+<p><strong>Activity: </strong></p>{{Activity}}
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.SendGridManagement/Emails/TaskDetail.html:5]: `{{Activity}}` inside the `<p>`?

File: CommunicationService/CommunicationService.SendGridManagement/Emails/TaskDetail.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+<p><strong>Organisation: </strong></p>{{Organisation}}
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.SendGridManagement/Emails/TaskDetail.html:6]: I think organisation is only populated when it's a face coverings request for an organisation...

File: CommunicationService/CommunicationService.SendGridManagement/Emails/TaskDetail.html [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    <tr><td>Shopping List</td></tr>
+    {{ShoppingList}}
+
+    <tr><td></td></tr>
+    <tr>Further Details</tr>
+    {{FurtherDetails}}
+
+</table>
+
+<p><strong>Volunteering Instructions</strong></p>
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.SendGridManagement/Emails/TaskDetail.html:20]: In the front end we call them "Volunteer Instructions" (singular)

File: CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+            {
+                throw new BadRequestException($"unable to retrieve user object for {recipientUserId.Value}");
+            }
+
+            var job = await _connectRequestService.GetJobDetailsAsync(jobId.Value);
+            var volunteerInstructions = "";
+            var group = await _connectGroupService.GetGroup(job.JobSummary.ReferringGroupID);
+           
+            if (group != null) {
+                var instructions = await _connectGroupService.GetGroupSupportActivityInstructions(groupId.Value, job.JobSummary.SupportActivity);
```
Comment (mike-f50) [inline CommunicationService/CommunicationService.MessageService/TaskDetailMessage.cs:80]: group.ID rather than groupID?

**Review by jawwadmukhtargmail — APPROVED** [review]
approved

**Pull request merged**